policy: Force L3 enforcement when toFQDNs is present via a selector

raybejjani opened the pull request:
When calculating L3/L4 policy, toFQDNs was ignored. In the cases where a
DNS lookup had not occured, and no toCIDRSet rules generated, the policy
was interpreted as L3 wildcard. We now treat toFQDNs as a label selector
style rule that always selects EntityNone. Nothing can create an
EntityNone so it should never be a valid target.

Fixes https://github.com/cilium/cilium/issues/6292

**update**
Here is a test I used:
```
$ cat tmp/empty-fqdn.yaml
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: rebel-escape
  namespace: default
spec:
  endpointSelector:
    matchLabels:
      any:org: alliance
  egress:
  - toEndpoints:
    - matchLabels:
       "k8s:io.kubernetes.pod.namespace": kube-system
       "k8s:k8s-app": kube-dns
    toPorts:
      - ports:
          - port: "53"
            protocol: ANY
        rules:
          dns:
            - matchPattern: "*cloudhmar.com"

  - toFQDNs:
      - matchName: "blsadsfadsfafds.cloudhmar.com"
      - matchName: "cloudhmar.com"
    toPorts:
      - ports:
        - port: "80"
```
* `$ kubectl exec xwing -- curl -svvv -4 -I cloudhmar.com --resolve cloudhmar.com:80:104.28.6.79` # timeout because no DNS lookup happens because of --resolve
* `$ kubectl exec xwing -- curl -svvv -4 -I cloudhmar.com --resolve cloudhmar.com:83:104.28.6.79` # success because a DNS lookup happens because --resolve doesn't apply here (port 83 not 80)
* `$ kubectl exec xwing -- curl -svvv -4 -I cname.cloudhmar.com --resolve cloudhmar.com:83:104.28.6.79` # timeout because the policy only allows connections to cloudhmar.com at L3, not any subdomains (except blsadsfadsfafds.cloudhmar.com but that doesn't exist)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6326)
<!-- Reviewable:end -->

raybejjani committed:
policy: Force L3 when toFQDNs is present via a selector

When calculating L3/L4 policy, toFQDNs was ignored. In the cases where a
DNS lookup had not occured, and no toCIDRSet rules generated, the policy
was interpreted as L3 wildcard. We now treat toFQDNs as a label selector
style rule that always selects EntityNone. Nothing can create an
EntityNone so it should never be a valid target.

Signed-off-by: Ray Bejjani <ray@covalent.io>

ianvernon commented on pkg/policy/api/fqdn.go:
```diff
@@ -83,3 +83,14 @@ func (r *PortRuleDNS) Sanitize() error {
 	return matchpattern.Validate(r.MatchPattern)
 }
+
+func (s *FQDNSelector) GetDestinationEndpointSelectors() EndpointSelectorSlice {
```
This function isn't used anywhere.

raybejjani replied:
yeah, I had started mimicking other rule types, but then ended up returning a single selector for the whole list (other types' `fooList.GetDestinationEndpointSelectors` iterated over the list and called `GetDestinationEndpointSelectors` on each).
I'll clean this up regardless.

ianvernon commented on pkg/policy/api/entity.go:
```diff
@@ -41,6 +41,8 @@ const (
 	// EntityInit is an entity that represents an initializing endpoint
 	EntityInit Entity = "init"
+
+	EntityNone Entity = "none"
```
Should we allow this entity to be provided via the API? I think we could allow policy which uses it, but not for labels themselves for endpoints. I.e., we could never provide an endpoint with labels `reserved:none`.

raybejjani replied:
That seems reasonable to me. One decision to make around this is whether this becomes the canonical way to express deny-all. It isn't critical to fixing the toFQDN bug, but introducing entityNone and then using `rule.toSomething == nil` and `len(rule.toSomething) = 0` to each mean  something might cause us trouble later.
Merging something like this PR allows the toGroups handling to change slightly, where it would be a deny-all when cilium-agent sees the parent policy, and the derivative policies whitelist the appropriate IPs. The current toGroups behaviour is to ignore the parent policy rules with toGroups outright, only beginning enforcement once a derivative policy is created.

eloycoto reviewed (approved):
```
[
  {
    "endpointSelector": {
      "matchLabels": {
        "container:id.app1": ""
      }
    },
    "egress": [
      {
        "toPorts": [
          {
            "ports": [
              {
                "port": "53",
                "protocol": "ANY"
              },
              {
                "port": "80",
                "protocol": "ANY"
              }
            ]
          }
        ],
        "toFQDNs": [
          {
            "matchName": "foo.foo.foo.foo.foo.com"
          }
        ]
      }
    ],
    "labels": [
      {
        "key": "toFQDNs-runtime-test-policy",
        "source": ""
      },
      {
        "key": "ToFQDN-UUID",
        "value": "7a7804ff-f488-11e8-87e7-080027d2d952",
        "source": "cilium-generated"
      }
    ]
  }
]
```

Testing working correctly. Thanks

raybejjani reviewed (commented):
@ianvernon We don't seem to have a RFC label, so I didn't want anyone to think this was final. I posted it up because we decided to do so in the SIG meeting. Amazingly, it passed tests!

aanm reviewed (approved):
What are the RFC for? this looks super good to me

raybejjani committed:
policy: Add EntityNone - a non-assignable entity

EntityNone expresses the opposite of EntityAll. It never matches
anything and is, in effect, an explicit deny-all.

Signed-off-by: Ray Bejjani <ray@covalent.io>

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1139,6 +1139,122 @@ var _ = Describe("RuntimePolicies", func() {
+		curlWithRetry(helpers.App1, "http://%s/public", http1IP[helpers.IPv4]).ExpectSuccess("%q cannot make http request to pod", helpers.App1)
+		curlWithRetry(helpers.App1, "-4 http://%s", worldIP).ExpectSuccess("%q cannot make http request to pod", helpers.App1)
+
+		By("testing egress toEntity: world in combination with previous rules (app1)")
```
You use `api.EntityAll` instead of `api.EntityWorld` in the below policy.

raybejjani replied:
Oops, good catch. One of these was supposed to be EntityWorld and the other EntityAll. I'll fix this up and make the `By` more correct.

ianvernon commented on test/runtime/Policies.go:
```diff
@@ -1139,6 +1139,122 @@ var _ = Describe("RuntimePolicies", func() {
+		curlWithRetry(helpers.App1, "http://%s/public", http1IP[helpers.IPv4]).ExpectSuccess("%q cannot make http request to pod", helpers.App1)
+		curlWithRetry(helpers.App1, "-4 http://%s", worldIP).ExpectSuccess("%q cannot make http request to pod", helpers.App1)
+
+		By("testing egress toEntity: world alone with EntityNone")
```
Same here.

ianvernon reviewed (commented):
Change looks good overall,  apart from the inconsistency between `api.EntityWorld` vs. `api.EntityAll`.
